### DISCUSSION TITLE
Depth texture in NPR Pipeline Screen Shader?

### DISCUSSION THREAD

**Discussion by Kolupsy** [bnpr/Malt#334] [Q&A]
Does the NPR Pipeline expose the depth texture by default / is there an 'easy' way to sample the depth in the screen shader?

**Comment by pragma37** [comment]
The easiest way to sample depth in screen shaders is simply reading the POSITION/surface_position().
If you need to sample other pixels you have the w channel of IN_NORMAL_DEPTH:
https://github.com/bnpr/Malt/blob/d43ed17fe1d45d396d5e79b08751b48b8a84bbfa/Malt/Pipelines/NPR_Pipeline/Shaders/NPR_ScreenShader.glsl#L18

**Reply by Kolupsy** [reply]
What am I missing here? I added this node setup to a screen graph and hooked that into the pipeline but its only white, even if I move very close or inside my objects 🤔 
![image](https://user-images.githubusercontent.com/88381426/174329822-3a4d087d-7a19-48a8-9810-425431a33034.png)

**Reply by pragma37** [reply]
I think you have 2 issues:
You are not using the "Layer Only" option, so the transparent layer is overriding everything else.
Depth is not linear, so you must be really close to a surface to see a darker tone:

![imagen](https://user-images.githubusercontent.com/36610999/174332503-ec5d60c7-b916-4249-8454-b246a57a3870.png)

**Reply by Kolupsy** [reply]
So does this mean the depth texture is only available inside the render layer graph, not inside the render graph too? I was able to get the result by following your instructions in the default render layer graph but there does not seem to be a 'layer only' option in the render graph

**Reply by Kolupsy** [reply]
Also immediate follow-up. Can I remap the depth to give me a proper unit distance so that I can compare the values from sampling the depth texture with actual blender distance values?

**Reply by pragma37** [reply]
> So does this mean the depth texture is only available inside the render layer graph, not inside the render graph too? 

Yes, and only if you plug the Scene socket (or plug the Normal Depth socket manually).

If you think about it, it doesn't make too much sense to have a default Depth in Render nodes because you would have to either ignore transparency or ignore anything behind transparency.

But you can of course plug the Normal Depth to a Render Layer custom output if you want.

> Can I remap the depth to give me a proper unit distance so that I can compare the values from sampling the depth texture with actual blender distance values?

You can use depth_to_z() or screen_to_camera().

**Comment by Kolupsy** [comment]
thanks, I was able to create some cool lens flare effects. I sample the depth texture at each light position to determine if it should be occluded. Works great! Now I would like to handle alpha as well. I guess this is the point where it gets really tricky but do I have access to an alpha texture (a texture that stores the maximum alpha at each pixel of the captured scene)? I think that would help with partially occluding the lens flares. 
I think ideally I would need a texture that stores the distance to all opaque surfaces in one channel and the max alpha in another channel. 
Probably not perfect but I think it should help cover a lot more cases. 
Does this sound like I need to spec into custom pipeline territory?
![image](https://user-images.githubusercontent.com/88381426/174489970-1996c1e9-c319-4e6d-9308-1436b4bfede8.png)

**Reply by pragma37** [reply]
Looks great!

> do I have access to an alpha texture (a texture that stores the maximum alpha at each pixel of the captured scene)?

That's just the Color output alpha in the RenderLayers node?

> I think ideally I would need a texture that stores the distance to all opaque surfaces in one channel and the max alpha in another channel.
Probably not perfect but I think it should help cover a lot more cases.
Does this sound like I need to spec into custom pipeline territory?

I'm not sure I get what you mean, but you can do pretty much everything you want with custom Python nodes.

**Comment by Kolupsy** [comment]
> That's just the Color output alpha in the RenderLayers node?

Haha you are absolutely right. Of course! Nice. This makes total sense. This made implementing transparancy possible and I was even able to throw in hue tinting ( only one layer though) into the mix. 

https://user-images.githubusercontent.com/88381426/174639555-69322546-e61a-4d43-96f1-6e2751cb81e4.mp4

The system works relatively well with sun lights and has some issues with point- and spot lights but I think those are inherit in the rasterized alpha handling.
